text/0122-sdk-hub-scope-merge.md

```diff
@@ -0,0 +1,148 @@
+- Start Date: 2023-11-07
+- RFC Type: feature
+- RFC PR: https://github.com/getsentry/rfcs/pull/122
+- RFC Status: draft
+
+# Summary
+
+We want to streamline Hub & Scope usage for the SDK.
+
+# Motivation
+
+Generally, Hubs and Scopes are a very Sentry-specific concept, and can be a bit hard to explain. 
+Also, how Hub & Scope forking behaves is currently a bit tricky, and does not play well e.g. with how OpenTelemetry contexts work.
+
+This RFC aims to streamline this by merging the concepts of Hub and Scope into a singular Scope concept.
+
+# Background
+
+TODO
+
+# Proposed Solution
+
+Hubs will be removed as a concept for SDK users. 
+Instead, from a users perspective only Scopes remain, which will become the singular entity to hold context data etc.
+
+Scopes will be _similar_ to how they work today, but not entirely the same. 
+Scopes can have data (e.g. tags, user, ...) added to them the same way you can do today. 
+This RFC _does not_ aim to change any of the data that is kept on the scope and is applied to events.
+
+The following APIs will be removed:
+
+* `getCurrentHub()`
+* `configureScope()` (instead just get the scope and set on it directly)
+* Any APIs currently on the Hub only: 
+  * `hub.pushScope()`
+  * `hub.popScope()`
+  * `hub.isOlderThan()`
+  * `hub.bindClient()`
+  * `hub.getStack()`
+  * `hub.getStackTop()`
+  * `hub.run()` (use `withScope()` instead)
+
+Instead, we will introduce some new APIs:
+
+```ts
+// get the currently active scope. replacement for `getCurrentHub().getScope()`
+export function getScope(): Scope;
+
+// get the currently active client. May return a NOOP client. Replacement for `getCurrentHub().getClient()`.
+export function getClient(): Client;
+
+// make a scope the current scope. Replacement for `makeMain(hub)`
+export function makeCurrentScope(scope: Scope): void;
+```
+
+The following APIs already exist but will behave differently:
+
+* `withScope()` will still work, but it will actually fork an execution context. So this will roughyl the same as doing `hub.run()` today.
+
+APIs that are currently on the hub should instead be called directly on the scope (e.g. `scope.captureException()` etc.), or via a global method (e.g. `Sentry.captureException()`).
+
+## Clients
+
+Instead of a client being optional, there will now _always_ be a client. It may be a Noop Client that does nothing, if `init()` has not been called yet.
+
+Each client gets a fresh initial scope when it is initialized:
+
+```js
+const client = new Client();
+client.getScope(); // <-- fresh, initial scope for this client
+scope.getClient(); // <-- get client of scope
```

Comment:
Aren't we introducing reference cycles with this?

Reply:
Yes, we are, I am not 100% sure if this is a problem 🤔 I _think_ at least in JS this should be fine as they should be garbage collectable if they are the only things referencing themselves (* I think). If this is problematic we can also leave the `scope.getClient()` API away I'd say - wdyt?

Reply:
I would leave this out of the scope of the RFC, and each platform can add it if they see the need.

Reply:
What's the reason for adding these APIs? As of now [clients are state-less](https://develop.sentry.dev/sdk/unified-api/#client), so IMHO it would make more sense to move the ownership into the scope instead. E.g. `Sentry.getScope().getClient()` and `Sentry.getScope().setClient()`.

Reply:
The idea was to make it easier to create (multiple) clients, but no strong feelings, we can also leave this on the scope only and require to do e.g. `scope.setClient(client)` or something like this. Then you'd do:

```js
const client1 = new Client();
const client2 = new Client();

const scope1 = new Scope();
const scope2 = new Scope();
scope1.setClient(client1);
scope2.setClient(client2);

scope1.captureException(...);
```

Which I guess is fine too!

Reply:
I updated this so only the scope has a reference to the client, but the client remains stateless.